### PR TITLE
Remove jQuery from sharing.js

### PR DESCRIPTION
**Pull request by sgomes**
Works towards #16409.

#### Changes proposed in this Pull Request:

* Reimplement `modules/sharedaddy/sharing.js` without jQuery.
* Remove dependency on `jQuery` when enqueuing

Note: I intentionally removed the slide animation on the dialogs. Slide animations are difficult to implement performantly on the web, and I felt like they didn't add enough here to justify the effort and the additional JS/CSS needed. If you feel it's important, however, please let me know and I'll be happy to go back and cook something up.

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Ensure you're testing the new code (e.g. through a sandbox or Chrome's local overrides)
* Enable sharing buttons in the Jetpack options for your test site
* Configure the sharing buttons so that there's at least one button inside a "More" dialog
* Configure the sharing buttons so that the "Print" and "Email" buttons are enabled
* Go to the test site and open a blog post
* Ensure that share counts are displayed on the buttons that support them (e.g. Facebook)
* Ensure that the More button displays the dialog on hover
* Ensure that the More button displays the dialog on touch / click
* Ensure that the Print button works
* Ensure that the Email button works
* Ensure that all of the above works on IE 11 as well
* If possible, also test the functionality on a site that lazily loads posts as the user scrolls (or clicks on something like a "More posts" button), and ensure that newly loaded posts have their share buttons working correctly too

#### Proposed changelog entry for your changes:

* Removed `jQuery` dependency from `modules/sharedaddy/sharing.js`

### CODE REVIEW

**Comment by sgomes** [comment]
Hey @gravityrail and @dero 👋 Any chance you could take a quick look at the PR, even if it isn't a full review + approval?

I'd love to get started on other scripts, but would like to validate the general approach first, if that's ok!

**Comment by gravityrail** [comment]
Sure! I'll take a look now 👍

File: modules/sharedaddy/sharing.js [hunk 1]
```diff
@@ -1,364 +1,474 @@
...
-							'//graph.facebook.com/?callback=WPCOMSharing.update_facebook_count&ids=' +
-							encodeURIComponent( url ),
-					],
-				};
-
-				for ( service in requests ) {
-					if ( ! jQuery( 'a[data-shared=sharing-' + service + '-' + id + ']' ).length ) {
-						continue;
-					}
+	// -------------------------- UTILITY FUNCTIONS -------------------------- //
```
Comment (gravityrail) [inline modules/sharedaddy/sharing.js:12]: As counter-intuitive as this sounds, I would like to create a separate PR (from which this PR is branched) which puts any shareable utility functions into a separate non-jQuery dependency. 

For example, in my carousel PR, I created a (badly-named, messy) library called `jpQuery` which included anything that was more than a 3-liner to replace.

https://github.com/Automattic/jetpack/pull/16420/files#diff-c2ea7f98aaf08faa387719e495ccff36

It seems to be that we are going to have a hard time ripping out jQuery without replacing _some_ non-trivial functions with a shared JS library, though hopefully one that's much smaller than jQuery.

What do you think of that idea? Would you be willing to create a PR that combines your utility functions with jpQuery and then rebase this PR on top of it? I would then do the same with my own PR.

File: modules/sharedaddy/sharing.js [hunk 2]
```diff
@@ -1,364 +1,474 @@
...
+	// Helper function to remove a node from the DOM.
+	function removeNode( node ) {
+		if ( node && node.parentNode ) {
+			node.parentNode.removeChild( node );
+		}
+	}
 
-				permalink = WPCOMSharing.get_permalink( url );
+	// Helper functions to show/hide a node, and check its status.
+	function hideNode( node ) {
```
Comment (gravityrail) [inline modules/sharedaddy/sharing.js:64]: I don't think this needs to be a separate function. It's more-or-less a one-liner in the code, so let's just do this inline.

File: modules/sharedaddy/sharing.js [hunk 3]
```diff
@@ -1,364 +1,474 @@
...
+		if ( node ) {
+			node.style.display = 'none';
+		}
+	}
 
-				if ( ! ( permalink in WPCOM_sharing_counts ) ) {
-					continue;
-				}
+	function showNode( node ) {
```
Comment (gravityrail) [inline modules/sharedaddy/sharing.js:70]: This also doesn't need to be a function when it's a one-liner.

File: modules/sharedaddy/sharing.js [hunk 4]
```diff
@@ -1,364 +1,474 @@
...
-		bump_sharing_count_stat: function ( service ) {
-			new Image().src =
-				document.location.protocol +
-				'//pixel.wp.com/g.gif?v=wpcom-no-pv&x_sharing-count-request=' +
-				service +
-				'&r=' +
-				Math.random();
-		},
+	function isNodeHidden( node ) {
+		return ! node || node.style.display === 'none';
```
Comment (gravityrail) [inline modules/sharedaddy/sharing.js:77]: We should just check this inline too.

File: modules/sharedaddy/sharing.js [hunk 5]
```diff
@@ -1,364 +1,474 @@
...
+	}
+
+	// ------------------------------- CLASSES ------------------------------- //
+
+	// Implements a MoreButton class, which controls the lifecycle and behavior
+	// of a "more" button and its dialog.
+	function MoreButton( buttonEl ) {
```
Comment (gravityrail) [inline modules/sharedaddy/sharing.js:84]: If we are going to introduce a prototype, we might as well introduce a class and things will get much simpler.

The hard thing here is IE11. Should we transpile _just_ for that browser?

One thing I'm thinking is... perhaps we do two builds, one which transpiles for evergreen browsers, and one for IE11. The evergreen browsers can use es6 classes natively, whereas IE11 gets the polyfill stuff. 

This will allow us to write much cleaner and more modern javascript, and the holdout can run the polyfills without everyone else getting them.

Might make a nice blog post at some point so that other plugins can emulate this approach.

File: modules/sharedaddy/sharing.js [hunk 6]
```diff
@@ -1,364 +1,474 @@
...
-							handler_item_leave = function () {
-								$more_sharing_button.data( 'hasitem', false );
+	// ---------------------------- SHARE COUNTS ---------------------------- //
 
-								if ( $more_sharing_button.data( 'hasoriginal' ) === false ) {
-									var timer = setTimeout( close_it, 800 );
-									$more_sharing_button.data( 'timer2', timer );
-								}
-							};
+	if ( window.sharing_js_options && window.sharing_js_options.counts ) {
```
Comment (gravityrail) [inline modules/sharedaddy/sharing.js:228]: Presumably `sharing_js_options` is set by `wp_localize_script`. If `sharing_js_options` is empty, then I _imagine_ we shouldn't load any of this JS at all. Let's move that condition into PHP so we don't load any JS that we aren't going to run.

File: modules/sharedaddy/sharing.js [hunk 7]
```diff
@@ -1,364 +1,474 @@
...
-						'" src="' +
-						urlToPrint +
-						'" onload="frames[\'printFrame-' +
-						uniqueId +
-						"'].focus();frames['printFrame-" +
-						uniqueId +
-						'\'].print();"></iframe>'
-				);
+				var iframe = document.createElement( 'iframe' );
+				iframe.style = 'position:fixed; top:100; left:100; height:1px; width:1px; border:none;';
```
Comment (gravityrail) [inline modules/sharedaddy/sharing.js:425]: Does this really work? I thought you had to set style to an object.

Comment (gravityrail) [reply modules/sharedaddy/sharing.js:425]: I think for a lot of these, you really want to be calling `iframe.setAttribute` rather than `iframe.attrname = 'foo'`. Don't know what the side effect of setting these as properties might be.

File: modules/sharedaddy/sharing.js (hunk 1 above)
Comment (dero) [reply modules/sharedaddy/sharing.js:12]: @gravityrail I fully agree we need some common base, I'm just not sure what the best mechanism to achieve it is.

These are some of the options I was considering:

1. We can either create a small common library like you suggest and enqueue it async / deferred,
2. or create / assemble (existing micro libraries) a collection of tools / methods and selectively `import` from it,
3. or we can introduce babel, write shiny ES6+ and inject a battery of polyfills for IE10+ under a conditional comment.

Actually, 2 + 3 could be combined. Also either one requires us adding an additional build step, but it could also lead to a clean code that doesn't depend on anything external. But it introduces complexity, potentially mainly around integration with existing build tools / linters / checks / wpcom.

Happy to follow any decision you'll make, just let me know and I can help out.

Comment (sgomes) [reply modules/sharedaddy/sharing.js:12]: Thank you, @dero and @gravityrail! I fully expected these sorts of concerns to come from an initial PR, where the approach hasn't been defined yet. For the record, I don't mind changing everything altogether or throwing out the code and starting anew; sometimes you have to try something before deciding it's not the right direction 😄 

So with that in mind, and taking other comments into mind here, I think @dero's three options above illustrate the possible paths we can take well.

Starting with option 1, this appears to be the easiest (but not necessarily best) path forward. We create a new library, enqueue it with WP's usual mechanisms, and use it where needed. Everything stays ES5 and IE11-compatible. The risk is that the library becomes a large monolith over time (probably never as big as jQuery, but still), with many different scripts depending on it. The upside is that even if that happens, we're in control of it, so we could e.g. at some point make it smaller by dropping IE11 compatibility, while maintaining the same external API.

Options 2 and 3 involve an entirely different approach, where code would be ES2015+ before being transpiled, and we could take advantage of modules for selective importing from a shared library (or two, or three), not to mention the much better developer experience. This approach breaks down into two further possibilities: either we ship a single ES5 IE11-compatible bundle to everyone, or we do differential serving based on the user's browser. Let's look at each of them separately:

**IE11-compatible bundle for everyone**

This is probably the easiest option of the two. We would enqueue a polyfill set and the transpiled bundle and call it a day. The downside is that the polyfills will almost certainly end up being larger than jQuery, which feels like running from one problem straight into another.

**Differential serving**

This one gets trickier, because as far as I know, we can't really rely on `enqueue` any more.

The general idea is that we'd prepare two bundles, one for old browsers and one for newer browsers, with an arbitrarily defined cutoff point, and we'd serve the right one to each browser.

That could be done with either a server-side mechanism (by looking at the user agent string and guessing which file to serve), or through a client-side mechanism, for example taking advantage of the module/nomodule trick (here's [an example article introducing the approach](https://dev.to/thejohnstew/differential-serving-3dkf)). I would recommend the latter for Jetpack.

Of course, this means we need to inject both a `<script nomodule src="..">` and a `<script type="module" src="...">` into the page, which AFAIK isn't really doable with `enqueue`, though I'm certainly not an expert there 🙂 

Note that even in this approach, we'd need to be vigilant about what we include, since if we're transpiling to this arbitrary point in time (in the case of `module`/`nomodule`, that would be the available feature set circa Chrome 61 / Safari 11), anything else would still need to be polyfilled. Still, it is an excellent path forward for maintaining compatibility on old browsers, improving performance on new ones, and doing all of this without major infrastructure changes to WP.

**Next steps**

So this looks like a key decision that needs to be made before any further work 😄 I'm happy to move this discussion elsewhere or schedule a call if you prefer, but I'm also happy to continue the discussion here. There are a number of possibilities as discussed above, and I'm happy to provide more details on all of them!

Comment (dero) [reply modules/sharedaddy/sharing.js:12]: > [...] battery of polyfills for IE10+ under a conditional comment.

After all these years and I still think conditional comments are a thing. #dinosaurInARoom 🦕 

Should we decide for transpilation, I agree with @sgomes that any bundle of polyfills need to be loaded selectively only on browsers that require them. I'm not a huge fan of the `module` / `nomodule` approach, however, because it's a hack.

I would rather do feature detection ([ES6](https://kangax.github.io/compat-table/es6/), [ES2016+](https://kangax.github.io/compat-table/es2016plus/)). With [Firefox implementing a new, ES2018 compatible RegExp engine in SpiderMonkey](https://www.mozilla.org/en-US/firefox/78.0/releasenotes/#note-788331), all current versions of modern browsers now basically support ES2019 and it shouldn't be too difficult to construct a compound check that only ES2019 (and lower) capable browsers satisfy and only load the polyfill bundle for older browsers. 

If that turns out not being possible for some reason, I wouldn't shy away from user-agent detection on the client. We're basically polyfilling only for Internet Explorer, all other browsers receive frequent updates and their feature set is quite comparable.

Comment (sgomes) [reply modules/sharedaddy/sharing.js:12]: I'm not a huge fan of the `module` / `nomodule` approach either, trust me 😄 Unfortunately, it's the only "free performance" option we have on the client.

Feature detection is certainly possible, but it adds a roundtrip cost to the scripts, which may or may not be acceptable. To clarify, with feature detection you need to wait for the feature detection script to load and run, determine which file is needed, and then wait for that file to be downloaded and run as well. This is perfectly acceptable for features like what this PR handles (share buttons), but may not be acceptable for any critical JS. Even for non-critical JS, it may make it hard to run in a specific sequence, if that's needed.

I'd shy away from server-side UA string matching if at all possible for this use-case. It requires extra server infrastructure that needs to be regularly updated, and UA strings will likely be getting frozen in the future, at least in Chrome.

File: modules/sharedaddy/sharing.js (hunk 5 above)
Comment (sgomes) [reply modules/sharedaddy/sharing.js:84]: (This discussion is continued on the first comment)

File: modules/sharedaddy/sharing.js (hunk 1 above)
Comment (dero) [reply modules/sharedaddy/sharing.js:12]: > Feature detection is certainly possible, but it adds a roundtrip cost to the scripts. [...] but may not be acceptable for any critical JS.

I understand the potential restrictions, @sgomes, and I agree with you that feature detection is not a magic silver bullet that would just work everywhere. But we may want to explore this path in more detail, nevertheless, because we don't _need_ it to work everywhere, we need to work in Jetpack and I'd argue we have a lot of knowledge on and control over the scripts that would be subject to these changes.

Here are a questions that may expand our options:

1. Is there a piece of JS in the Jetpack codebase that would be "critical" enough that it couldn't be delayed? (Any inline JS excluded, those are likely going to remain ES5 compatible.)
2. Are the pieces of JS we're stripping jQuery from interdependent? They usually live in modules that can be enabled independently, so I'm inclined to say "mostly no".
3. Is there any external JS Jetpack runs that needs to run synchronously? We have been working on improving other aspects of Jetpack's performance earlier this year and I don't recall that being the case.

Can we make the ES6 a default that falls back to IE11-compat _on the client_? Maybe something like this could work? It's a bit... let's say _experimental_, but maybe worth a discussion:

```js
// Top of the page. Could be further optimized to init Jetpack_JS async if necessary.
var Jetpack_JS = {
  queue: function( name, fn ) {
    // Runs and memoizes a test checking the current browser ES version capabilities.
    // If the test fails, `loadScript`s an `es5` version of the JS and overrides the fn. Honors queue order during execution.
    // Otherwise run the original ES2019 code immediately.
  };
};

// In (for example) tiled-gallery.js.
Jetpack_JS.queue( 'tiled-gallery', function() {
  // ES2019 code.
} );
```

This approach would simply run the ES2019 code in modern browsers with a (probably) negligible overhead.

IE11, Blink Edge and any other outdated browsers we may wish to support would suffer a performance hit and would end up double-downloading the JS files, but I'd argue that's not much of an issue, because those are desktop browsers and we're not executing the scripts twice, only downloading. And I suppose our end goal is to _support_ IE11 and Blink Edge functionally, but not necessarily optimize for performance in those outdated browsers.

Comment (gravityrail) [reply modules/sharedaddy/sharing.js:12]: Thanks for the really robust discussion. It helps me a lot to read the pros and cons of each approach. 

Given the options described above, here's what I'd recommend:

- Let's use a build system and transpile everything.
- We really only need to detect one user-agent, which is IE11. We will serve "modern" JS to every other browser. Maybe that browser can't run it, but I'm prepared to wear that cost for the sake of simplicity.
- We build our jquery-replacement utility library (let's call it 'common' for now) as an npm package inside the Jetpack monorepo, and import things like this: `import { forEachNode } from '@jetpack/common'`
- As we convert JS to remove jQuery, we _also_ migrate it to this new transpilation system, which does bundle-splitting so that functions that are common across all features end up in a `vendor.min.js` file of some kind - **PLUS** we do this **TWICE**, once each for the two browser-specific builds I will describe next...
- For IE11, we serve a transpiled ES5 + polyfills bundle
- For **everyone** else, we serve a bundle optimized for the last 2 major versions of Chrome, Safari, Firefox and Edge

So, basically, that's it. Since IE11 is the only holdout on nice things like ES6 classes, and since IE11 needs the vast majority of polyfills, this should mean that all the compatibility garbage is reserved for IE11 users and everyone else gets a much more lean build.

Also, in terms of browser detection, we really _only_ need to do that for IE11. If it introduces a performance penalty, I don't really care since IE11 support will go away at some point anyway, at which point we just delete that config from the build system.

What do you think of that?

Comment (sgomes) [reply modules/sharedaddy/sharing.js:12]: Thank you both, @dero and @gravityrail! Now that I understand Jetpack's needs a bit better I agree that feature detection sounds like a good idea, and I especially like your idea of zero-cost green path with fallback, @dero! That's a clever approach that I'd definitely love to try 👍 

The baseline of "last 2 major versions of Chrome, Safari, Firefox and Edge" that @gravityrail suggested seems perfect. Whenever possible, we should probably try not to use the absolute latest bells and whistles, but we can be reasonably confident that the vast majority of users will be on those releases.

Taking all of the discussion into account, here's how I'd put everything together:
- Perform a test at the beginning of the script, to see if the browser supports basic ES2015 syntax and any newer features that the script makes use of. If not, go to fallback approach. Crucially, this should happen before the script makes any changes to the page, so we don't have to worry about rolling things back or anything.
- The fallback approach is to stop execution immediately and load the fallback script asynchronously. In order for this to work, we need to ensure that all Jetpack scripts work independently of each other, and in any loading order. I think this might already be the case, right?
- Anything that passes the test for "modern browser" but fails execution later is considered a bug, or a side effect of this approach, which we may or may not want to fix. Reducing the amount of users that fall into this category is why I would suggest testing for newer features as well, instead of just basic ES2015 syntax. Of course, we can make these tests part of `@jetpack/common` to make things easier.

As for how to handle the polyfills, we have two options. The difficult, but more performant option, is to have a separate polyfill script that we load and run before the fallback scripts. Here we run into the trickiness of having to manage execution order.

The much easier option, and the one I'd go for, is to include the polyfills with every Jetpack fallback script, which of course means a lot of duplication if there are a bunch of Jetpack options enabled, but significantly simplifies things on our end. And echoing @gravityrail's approach, I think that's a perfectly acceptable tradeoff; we probably shouldn't spend much time thinking about IE11 perf, and as long as things work at a basic level, that's ok. If users want better performance, they'll need to upgrade their browser.

I'll get started by working on a new PR that reimplements this script for modern browsers; it's going to feel great to rewrite the whole thing to be 2x smaller 😄  Once I'm done with that, we can see which bits seem to be reusable, so that we can pull them out into the utility library that @gravityrail suggested, and we can see about adding the feature-detection bits and coming up with that build process! Happy for someone else to look into that in the meantime too, if they're interested 👍

**Comment by sgomes** [comment]
As per the discussion in https://github.com/Automattic/jetpack/pull/16592#discussion_r463200046, this PR will likely not be moving forward as it is. I'm keeping it open for now, but will start work on a new version following the discussed approach.

File: modules/sharedaddy/sharing.js (hunk 2 above)
Comment (sgomes) [reply modules/sharedaddy/sharing.js:64]: Yeah, I wrote it this way, but I agree it's somewhat debatable. I did it more to make things a bit more semantically meaningful; code is a bit easier to read if you see `hideNode` rather than setting a property on a node's `style`, but I agree it may not be saving much. It also helps a bit if we later opt for a different hiding approach (say, with a fadeout), but that doesn't seem like something we'd go for anytime soon.

I'm happy to go either way here.

File: modules/sharedaddy/sharing.js (hunk 6 above)
Comment (sgomes) [reply modules/sharedaddy/sharing.js:228]: The way the script was written, it makes me think that `sharing_js_options` is only needed for the sharing counts, and not for the rest of the sharing button functionality that the script handles (more button, print button, email button, etc). I don't really know if that's the case, though, or if it's just a coding mistake in the first place.

File: modules/sharedaddy/sharing.js (hunk 7 above)
Comment (sgomes) [reply modules/sharedaddy/sharing.js:425]: I'll double-check all of this in a few browsers 👍

**Comment by sgomes** [comment]
Added comments to all of the other suggestions I'd left on hold while we were discussing how to move forward.

File: modules/sharedaddy/sharing.js (hunk 3 above)
Comment (sgomes) [reply modules/sharedaddy/sharing.js:70]: Same as above.

File: modules/sharedaddy/sharing.js (hunk 4 above)
Comment (sgomes) [reply modules/sharedaddy/sharing.js:77]: Same as above.

File: modules/sharedaddy/sharing.js (hunk 7 above)
Comment (sgomes) [reply modules/sharedaddy/sharing.js:425]: Note: this seems to work in every browser I've tried. What we're doing here is setting the "style" property on the iframe element. Since DOM properties are strings, this works exactly as intended.

Comment (sgomes) [reply modules/sharedaddy/sharing.js:425]: Reading over the spec, it's probably best to set this as an attribute, though. The property is supposed to be an object, and the attribute is supposed to be a string. The fact that setting the property to a string works is probably reliable and probably spec-compliant, but it's not the best way of doing things.

File: modules/sharedaddy/sharing.js (hunk 1 above)
Comment (gravityrail) [reply modules/sharedaddy/sharing.js:12]: Thank you so much! This feels like a good direction. I really appreciate all your efforts and how you're incorporating feedback into the approach.

**Comment by sgomes** [comment]
Note: this PR is currently stale.
